### ISSUE TITLE
missing CallExpression following NewExpression

### ISSUE THREAD

**Issue by ghissues** [jquery/esprima#1306]
## new function(){17}()
    to simplify the parse i use a Literal: 17
    for testing i replace the literal with a console.log("executed and called") call

(i) new function(){17}()
(ii) function(){17}()
(iii) f=function(){17}()
(iv) f= new function(){17}()
#### (iv) executes but omits || skips || misses parsing the CallExpression
results when passed through esprima:

# (i)
#### note:
    unexpected, omitted CallExpression in parse
#### source:
    new function(){17}()
#### with console.log() call console displays:
    executed and called
#### [esprima returns](http://esprima.org/demo/parse?code=new%20function()%7B17%7D()%0A)
    {
    "type": "Program",
    "body": [
        {
            "type": "ExpressionStatement",
            "expression": {
                "type": "NewExpression",
                "callee": {
                    "type": "FunctionExpression",
                    "id": null,
                    "params": [],
                    "defaults": [],
                    "body": {
                        "type": "BlockStatement",
                        "body": [
                            {
                                "type": "ExpressionStatement",
                                "expression": {
                                    "type": "Literal",
                                    "value": 17,
                                    "raw": "17"
                                }
                            }
                        ]
                    },
                    "generator": false,
                    "expression": false
                },
                "arguments": []
            }
        }
    ]
    }

# (ii)
#### note:
    expected, fails execution and parsing
#### source:
    function(){17}()
#### with console.log() call console displays:
    Uncaught SyntaxError: Unexpected token (
#### [esprima returns](http://esprima.org/demo/parse?code=function()%7B17%7D()%0A)
    Error: Line 1: Unexpected token (

# (iii)
#### note:
    expected, executes and parses
#### source:
    f=function(){17}()
#### with console.log() call console displays:
    executed and called
#### [esprima returns](http://esprima.org/demo/parse?code=f%3Dfunction()%7B17%7D()%0A)
    {
    "type": "Program",
    "body": [
        {
            "type": "ExpressionStatement",
            "expression": {
                "type": "AssignmentExpression",
                "operator": "=",
                "left": {
                    "type": "Identifier",
                    "name": "f"
                },
                "right": {
                    "type": "CallExpression",
                    "callee": {
                        "type": "FunctionExpression",
                        "id": null,
                        "params": [],
                        "defaults": [],
                        "body": {
                            "type": "BlockStatement",
                            "body": [
                                {
                                    "type": "ExpressionStatement",
                                    "expression": {
                                        "type": "Literal",
                                        "value": 17,
                                        "raw": "17"
                                    }
                                }
                            ]
                        },
                        "generator": false,
                        "expression": false
                    },
                    "arguments": []
                }
            }
        }
    ]
    }

# (iv)
#### note:
    unexpected, omitted CallExpression in parse
#### source:
    f=new function(){17}()
#### with console.log() call console displays:
    executed and called
#### [esprima returns](http://esprima.org/demo/parse?code=f%3Dnew%20function()%7B17%7D()%0A)
    {
    "type": "Program",
    "body": [
        {
            "type": "ExpressionStatement",
            "expression": {
                "type": "AssignmentExpression",
                "operator": "=",
                "left": {
                    "type": "Identifier",
                    "name": "f"
                },
                "right": {
                    "type": "NewExpression",
                    "callee": {
                        "type": "FunctionExpression",
                        "id": null,
                        "params": [],
                        "defaults": [],
                        "body": {
                            "type": "BlockStatement",
                            "body": [
                                {
                                    "type": "ExpressionStatement",
                                    "expression": {
                                        "type": "Literal",
                                        "value": 17,
                                        "raw": "17"
                                    }
                                }
                            ]
                        },
                        "generator": false,
                        "expression": false
                    },
                    "arguments": []
                }
            }
        }
    ]
    }

**Comment by michaelficarra** [comment]
These parses are all fine.

**Comment by ghissues** [comment]
my bad, can you explain the reasoning for the missing CallExpression?
esprima's docs are light on content and even reading through the SpiderMonkey docs all i could find was:

    interface NewExpression <: Expression {
        type: "NewExpression";
        callee: Expression;
        arguments: [ Expression ];
    }
    A new expression.
    
    interface CallExpression <: Expression {
        type: "CallExpression";
        callee: Expression;
        arguments: [ Expression ];
    }
    A function or method call expression.

and since the code calls the functions in (i),(iii),(iv) 
why is the CallExpression present in the parse in (iii) but omitted from (i) and (iv)?

**Comment by michaelficarra** [comment]
Because (i) and (iv) aren't function calls. The parens belong to the `NewExpression`.

**Comment by ghissues** [comment]
i dig, but there is still a call present so i wonder why leave it out of the AST?

looked up new operator on MDN:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new

    When the code new Foo(...) is executed, the following things happen:

    1. A new object is created, inheriting from Foo.prototype.
    2. The constructor function Foo is called with the specified 
    arguments and this bound to the newly created object. 
    new Foo is equivalent to new Foo(), i.e. if no argument 
    list is specified, Foo is called without arguments.
    3. The object returned by the constructor function becomes 
    the result of the whole new expression. 
    If the constructor function doesn't explicitly return an object, 
    the object created in step 1 is used instead. 
    (Normally constructors don't return a value, but they can choose 
    to do so if they want to override the normal object creation process.)

##### Foo is called

so the CallExpression should be understood as 'implied' in every NewExpression
to incorporate a CallExpression in the AST would be redundant?

**Comment by michaelficarra** [comment]
Foo is constructed. There is no call. Also, the AST represents syntactic constructs, not the semantics.

**Comment by ghissues** [comment]
i can appreciate semantics vs syntax

but i think that at least some semantics are important in correctly parsing syntax

or at least in identifying errors

when you said:

    The parens belong to the NewExpression.

i am assuming you meant the trailing paren pair ()?

but if i put them on a string constuctor:

    new String(17)()

my console gives me an error:

    Uncaught TypeError: object is not a function

implying the () belong to the function stead the NewExpression

but esprima parses without error.. and includes a CallExpression:

http://esprima.org/demo/parse?code=new%20String(1)()%0A

    {
    "type": "Program",
    "body": [
        {
            "type": "ExpressionStatement",
            "expression": {
                "type": "CallExpression",
                "callee": {
                    "type": "NewExpression",
                    "callee": {
                        "type": "Identifier",
                        "name": "String"
                    },
                    "arguments": [
                        {
                            "type": "Literal",
                            "value": 1,
                            "raw": "1"
                        }
                    ]
                },
                "arguments": []
            }
        }
    ]
    }

how does one explain this parsing?

**Comment by fkling** [comment]
Your last example, `new String(17)()` is indeed indeed a CallExpression + NewExpression. The first set of parens (`(17)`) belongs to the NewExpression, the second set to the CallExpression. It's evaluated as

```
new String(17)()
^------------^
    NewE
^--------------^
      CallE
```

However, note that this is **very** different from your original example:

````js
new function(){17}()
```

Here we only have a NewExpression. There is no second set of parens. If you wanted to adjust this to fit your last example, the equivalent would be

```
new function(){17}()()
^------------------^
    NewE
^--------------------^
      CallE
```

----

To recap: NewExpressions and CallExpressions may **look** similar, but they are different constructs. In particular, without arguments, the parens are *optional* in a NewExpression, i.e. `new Foo` or `new Foo()` are equivalent. However, the they are *not* optional in a CallExpression. `foo` and `foo()` are two different things.

**Comment by ghissues** [comment]
ok, i see how in this case esprima is doing the parsing

it seems to me inconsistent 

when building alongside the AST if i am interested in calls in the code i have to grok all CallExpressions and NewExpressions stead just the calls if it was consistently 

        "type": "ExpressionStatement",
        "expression": {
            "type": "CallExpression",
            "callee": {
                "type": "NewExpression",
                "callee": {},
                "arguments": []
            },
            "arguments": []
        }
   

for all NewExpressions

again, MDN uses **call** in its language when talking about these things

https://developer.mozilla.org/en-US/docs/Glossary/Constructor

    To call the constructor of the class in JavaScript, 
    use a new operator to assign a new object reference to a variable.



furthermore.. still confused why **new String(17)()** parses while it errors in my vm

one can claim esprima is only interested in syntax, but esprima identifies other errors why ignore this one?

**Comment by fkling** [comment]
> again, MDN uses call in its language when talking about these things

MDN is a public, non-normative wiki. It uses language that is easy to understand, but not necessarily technically correct.

The [official spec][es6] clearly distinguishes between `new` and function calls: The [`new` operator][new] invokes a function's internal `[[Construct]]` method while a [function call][] invokes a function's internal `[[Call]]` method.

Note that this is not the reason why esprima makes this distinction, but it shows that it is not arbitrary.

More importantly though, the spec defines these production rules:

*`new` operator*
```
NewExpression : new NewExpression
MemberExpression : new MemberExpression Arguments
```

*function calls*

```
CallExpression : MemberExpression Arguments
```

As you can see, there is a clear distinction between using `Foo()` and `new Foo()` in the spec. The production rules look similar but the ones for the `new` operator don't contain CallExpression at all.

The AST that *you* expected would assume a production rule like

```
MemberExpression : new CallExpression
```

but such a rule does **does not exist**.

Hence esprima is very close to spec in this case which is a good thing.

> one can claim esprima is only interested in syntax, but esprima identifies other errors why ignore this one?

There are so called *early errors* which are somehow between syntax errors and runtime errors. They require some minimal form of static analysis and the spec requires (I think) that those are raised at parse time, so it makes for esprima to error in these cases.

Anything else though is out of scope for esprima.

Here is an even simpler example:

```js
require
```

Would you expect esprima to throw a parse time error? If you are true to your position and considering your example, then you should say *yes*.

This would be a runtime error in the browser, because an identifer `require` does not exist. 
*However* this is a perfectly valid program in Node.js, where `require` exists. If esprima threw an error, it would do so incorrectly if the code is supposed to run in Node.js (which esprima does not know).

Similarly for your example, the code may run in an environment where `new String()` returns a function.

[es6]: http://www.ecma-international.org/ecma-262/6.0/index.html
[new]: http://www.ecma-international.org/ecma-262/6.0/index.html#sec-new-operator
[function call]: http://www.ecma-international.org/ecma-262/6.0/index.html#sec-function-calls

**Comment by michaelficarra** [comment]
This is not the place to learn about JavaScript grammar. We've identified that there is no problem, and this issue should be closed. Please move further discussion to the [mailing list](https://groups.google.com/forum/#!forum/esprima).
